lib/java/src/org/apache/thrift/utils/StringUtils.java

```diff
@@ -49,12 +49,10 @@ public static String bytesToHexString(byte[] bytes) {
    * @return hex string.
    */
   public static String bytesToHexString(byte[] bytes, int offset, int length) {
-    if (length < 0) {
-      throw new IllegalArgumentException("Negative length " + length);
-    }
-    if (offset < 0) {
-      throw new IndexOutOfBoundsException("Negative start offset " + offset);
+    if ((offset | length | (offset + length) | (bytes.length - (offset + length))) < 0) {
```

Comment:
To be honest,I personally have a problem with code like this where you need to read it three times to get it right (and I'm still not fully sure I "groked" it). I leave it to someone else to give a final review statement. +0 from me stince I won't be standing in the way, but I'm not really in favor of it.

Reply:
A testcase for this change would be a great addition.

Reply:
```suggestion
    if (offset < 0 || bytes.length < (offset + length)) {
```

And maybe we should keep the IllegalArgumentException when length is less than 0?

Reply:
I'm responding to the request on the mailing list for another pair of eyes on this PR.

I agree with @Jens-G regarding the readability... readability is far more important than an every-so-slight performance benefit. Obfuscated code is very hard to maintain, even with good test cases.

@joaopedroantonio 's suggestions are a great improvement, and preserves existing behavior, while adding the bounds checks when using offset and length.

A unit test case that covers the edge cases would be a nice sanity check for this method.

Reply:
> if ((offset | length | (offset + length) | (bytes.length - (offset + length))) < 0)

when length is less than 0, it will throw IllegalArgumentException, such as `int offset=2,length=-1,byteLength=10;`

Reply:
it seem that there will always be readability problem when using bit operation.

I do agree with that the readability of code is important. The code will be changed and squashed to a single commit.

Thanks very much.

Reply:
Maybe I'm being picky here but I think we should keep the IllegalArgumentException when `length < 0`, for two reasons:
1. The semantics of IndexOutOfBoundsException and the IllegalArgumentException are different and I think `length < 0` is a case of the latter.
2. Neither of those exceptions is checked and changing the type of exception when `length < 0` will not be explicit in compile time. I'd consider this a breaking change in the API.

WYT?

Reply:
I haven't notice that I have removed IllegalArgumentException instead of IndexOutOfBoundsException, and I haven't notice that it will be a breaking change in the API.

great suggestion, thanks very much.